File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.py
import sys
from collections import deque

graph = [0 for _ in range(101)]
visited = [False for _ in range(101)]

def bfs():
    queue = deque([(1, 0)])
    visited[1] = True

    while queue:
        pos, c = queue.popleft()

        if pos == 100:
            print(c)
            break

        for i in range(1, 7):
            pos_n = pos + i

            if 0 < pos_n <= 100:
                if graph[pos_n] != 0:
                    pos_n = graph[pos_n]

                if not visited[pos_n]:
                    visited[pos_n] = True
                    queue.append((pos_n, c + 1))

n, m = map(int, sys.stdin.readline().strip().split())

for i in range(n):
    j, k = map(int, sys.stdin.readline().strip().split())
    graph[j] = k

for i in range(m):
    j, k = map(int, sys.stdin.readline().strip().split())
    graph[j] = k

bfs()

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.cc
#include <iostream>
#include <vector>
#include <deque>

using namespace std;

void bfs(const vector<int>& graph) {
    deque<int> queue;
    vector<int> visited(101);

    queue.emplace_back(1);

    while (!queue.empty()) {
        const int node = queue.front();
        queue.pop_front();

        if (node == 100) {
            cout << visited[100] << "\n";
            return;
        }

        for (int i = 1; i <= 6; i++) {
            int nextNode = node + i;

            if (nextNode <= 100 && !visited[nextNode]) {
                if (graph[nextNode]) {
                    if (!visited[graph[nextNode]]) {
                        nextNode = graph[nextNode];
                        visited[nextNode] = visited[node] + 1;
                        queue.emplace_back(nextNode);
                    }
                } else if (!visited[nextNode]) {
                    visited[nextNode] = visited[node] + 1;
                    queue.emplace_back(nextNode);
                }
            }
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n, m;
    cin >> n >> m;

    vector<int> graph(101);

    for (int i = 0; i < n; i++) {
        int x, y;
        cin >> x >> y;

        graph[x] = y;
    }

    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;

        graph[u] = v;
    }

    bfs(graph);

    return 0;
}
